rust/src/lib/nm/connection.rs

```diff
@@ -251,39 +213,55 @@ pub(crate) fn get_port_nm_conns<'a>(
     ret
 }
 
-pub(crate) fn gen_nm_connection(
-    iface_name: &str,
-    uuid: &str,
-    nm_iface_type: &str,
-    ctrl_name: Option<&str>,
-    nm_ctrl_type: Option<&str>,
-    is_controller: bool,
-) -> NmConnection {
-    let mut nm_conn = NmConnection::new();
-
-    // OVS port already has it own prefix
-    let conn_name = if nm_iface_type == "ovs-bridge" {
-        format!("ovs-br-{}", iface_name)
-    } else if nm_iface_type == "ovs-interface" {
-        format!("ovs-iface-{}", iface_name)
+pub(crate) fn gen_nm_conn_setting(
+    iface: &Interface,
+    nm_conn: &mut NmConnection,
+) -> Result<(), NmstateError> {
+    let mut nm_conn_set = if let Some(cur_nm_conn_set) = &nm_conn.connection {
+        cur_nm_conn_set.clone()
     } else {
-        iface_name.to_string()
+        let mut new_nm_conn_set = NmSettingConnection::new();
+        let conn_name = if iface.iface_type() == InterfaceType::OvsBridge {
+            format!("ovs-br-{}", iface.name())
+        } else if iface.iface_type()
+            == InterfaceType::Other("ovs-port".to_string())
+        {
+            format!("ovs-port-{}", iface.name())
+        } else if iface.iface_type() == InterfaceType::OvsInterface {
+            format!("ovs-iface-{}", iface.name())
+        } else {
+            iface.name().to_string()
+        };
+
+        new_nm_conn_set.id = Some(conn_name);
+        new_nm_conn_set.uuid = Some(NmApi::uuid_gen());
+        new_nm_conn_set.iface_type =
+            Some(iface_type_to_nm(&iface.iface_type())?);
+        new_nm_conn_set
     };
 
-    let mut nm_conn_set = NmSettingConnection::new();
-    nm_conn_set.id = Some(conn_name);
-    nm_conn_set.uuid = Some(uuid.to_string());
-    nm_conn_set.iface_type = Some(nm_iface_type.to_string());
-    nm_conn_set.iface_name = Some(iface_name.to_string());
+    nm_conn_set.iface_name = Some(iface.name().to_string());
     nm_conn_set.autoconnect = Some(true);
-    nm_conn_set.autoconnect_ports =
-        if is_controller { Some(true) } else { None };
+    nm_conn_set.autoconnect_ports = if iface.is_controller() {
+        Some(true)
+    } else {
+        None
+    };
 
+    let nm_ctrl_type =
```

Comment:
Can be:
```
let nm_ctrl_type = iface
        .base_iface()
        .controller_type
        .as_ref()
        .map(iface_type_to_nm)
        .transpose()?;
```

Reply:
Why is the transpose required here? Probably I am missing something..

Reply:
Why is the trnspose() required? If I am not wrong, this should return only one element.. Am I missing something? Thanks!

Reply:
The `map()` here makes `Option<Result<T>>`, so doing transpose 'flips' it into `Result<Option<T>>` on which you can use `?`.